compiler/qsc_partial_eval/src/lib.rs

```diff
@@ -367,6 +372,210 @@ impl<'a> PartialEvaluator<'a> {
         Ok(EvalControlFlow::Continue(array_value))
     }
 
+    fn eval_bin_op_with_lhs_result_operand(
+        &mut self,
+        bin_op: BinOp,
+        lhs_result: val::Result,
+        rhs_expr_id: ExprId,
+        bin_op_expr_span: Span, // For diagnostic purposes only.
+    ) -> Result<EvalControlFlow, Error> {
+        let rhs_control_flow = self.try_eval_expr(rhs_expr_id)?;
+        let EvalControlFlow::Continue(rhs_value) = rhs_control_flow else {
+            let rhs_expr = self.get_expr(rhs_expr_id);
+            return Err(Error::Unexpected(
+                "embedded return in RHS expression".to_string(),
+                rhs_expr.span,
+            ));
+        };
+        let Value::Result(rhs_result) = rhs_value else {
+            panic!("expected result value from RHS expression");
+        };
+
+        // Get the operands to use when generating the binary operation instruction.
+        let lhs_operand = self.eval_result_as_bool_operand(lhs_result);
+        let rhs_operand = self.eval_result_as_bool_operand(rhs_result);
+
+        // Create a variable to store the result of the expression.
+        let variable_id = self.resource_manager.next_var();
+        let rir_variable = rir::Variable {
+            variable_id,
+            ty: rir::Ty::Boolean, // Binary operations between results are always Boolean.
+        };
+
+        // Create the binary operation instruction and add it to the current block.
+        let condition_code = match bin_op {
+            BinOp::Eq => ConditionCode::Eq,
+            BinOp::Neq => ConditionCode::Ne,
+            _ => {
+                return Err(Error::Unexpected(
+                    format!("invalid binary operator for Result operands: {bin_op:?})"),
+                    bin_op_expr_span,
+                ))
+            }
+        };
+        let instruction = Instruction::Icmp(condition_code, lhs_operand, rhs_operand, rir_variable);
+        self.get_current_rir_block_mut().0.push(instruction);
+
+        // Return the variable as a value.
+        let value = Value::Var(map_rir_var_to_eval_var(rir_variable));
+        Ok(EvalControlFlow::Continue(value))
+    }
+
+    fn eval_bin_op_with_lhs_classical_bool_operand(
+        &mut self,
+        bin_op: BinOp,
+        lhs_bool: bool,
+        rhs_expr_id: ExprId,
+    ) -> Result<EvalControlFlow, Error> {
+        let value = match (bin_op, lhs_bool) {
+            // Handle short-circuiting for logical AND and logical OR.
+            (BinOp::AndL, false) => Value::Bool(false),
+            (BinOp::OrL, true) => Value::Bool(true),
+            // The other possible cases.
+            (BinOp::AndL | BinOp::OrL | BinOp::Eq | BinOp::Neq, _) => {
```

Comment:
I think a similar thing can happen here... if the case is `(BinOp::AndL, true)` then the result is just the rhs, so rather than emitting a comparison it can be a store of rhs into the rir variable. Same for `(BinOp::OrL, false)`, `(BinOp::Eq, true)` and `(BinOp::Neq, false)`.

Reply:
Updated :)